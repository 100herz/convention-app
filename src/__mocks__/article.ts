import { Article } from '@models/article'

export const article: Article = {
  id: 18569,
  date_gmt: '2020-11-11T08:25:41',
  title: {
    rendered: '„New Normal“ bei Leonardo Hotels &#8211; mit neuen MICE-Formaten und Konzepten',
  },
  content: {
    rendered:
      '<p>Durch smarte Raumanpassungen, große Flexibilität bei Buchungen und Stornierungen sowie die Etablierung eines extern zertifizierten und weit über den Standard hinaus gehenden Sicherheits- und Hygienekonzeptes hat sich Leonardo Hotels umfassend auf das „New Normal“ und die geänderten Anforderungen ihrer MICE-Kunden an zukunftsfähigen Veranstaltungsformaten eingestellt. Neue Pausenkonzepte, digitalen Helfer, verlässliche Partner sowie virtuelle und hybride Veranstaltungsformate umfassen die Angebotsvielfalt der Leonardo Hotels.  Diese sind überzeugt, dass persönliche Meetings mit Sicherheit wieder Zukunft haben.</p>\n',
    protected: false,
  },
  excerpt: {
    rendered:
      '<p>Durch neue Sicherheits- und Hygienekonzeptes hat sich Leonardo Hotels umfassend auf das „New Normal“ und die geänderten Anforderungen an zukunftsfähigen Veranstaltungsformaten eingestellt.</p>\n',
    protected: false,
  },
  featured_media: 18571,
  categories: [6, 1],
  acf: {
    sponsored_by: 'XYZ',
    should_push: false,
    pinned: false,
    featured_slider: false,
    news_slider: true,
  },
  featured_image_thumb:
    'https://www.convention-net.de/wp-content/uploads/2020/11/iStock-1250329916-Personen_mit_Masken-w1500-h1300-300x300.jpg',
  featured_image_medium:
    'https://www.convention-net.de/wp-content/uploads/2020/11/iStock-1250329916-Personen_mit_Masken-w1500-h1300-600x400.jpg',
  categories_names: ['Hotels', 'News'],
}

export const articleWithoutFeatured: Article = {
  id: 18592,
  date_gmt: '2020-11-10T11:00:19',
  title: {
    rendered:
      'Events and Hospitality Industries Urged to Harness the Power of Digital Transformation to Thrive in 2021 and Beyond',
  },
  content: {
    rendered:
      '<p><a href="https://bit.ly/3iPkUhB">Cvent</a> CEO and Founder, Reggie Aggarwal, delivered an inspiring keynote to more than 7,000 event, marketing, and hospitality professionals at <a href="https://bit.ly/30Unj4A">Cvent CONNECT Europe</a> today. In his opening presentation, Aggarwal urged attendees of the virtual event to seize the digital transformation opportunity in front of them to drive the industry’s recovery in 2021 and beyond.</p>\n',
    protected: false,
  },
  excerpt: {
    rendered:
      '<p>Cvent CEO Reggie Aggarwal predicts “Golden Age of Events” during his keynote to more than 7,000 virtual attendees at Cvent CONNECT Europe </p>\n',
    protected: false,
  },
  featured_media: null,
  categories: [358, 217, 1],
  acf: {
    sponsored_by: '',
    should_push: false,
    pinned: false,
    featured_slider: false,
    news_slider: true,
  },
  featured_image_thumb: '',
  featured_image_medium: '',
  categories_names: ['Branche', 'Events', 'News'],
}

export const articleWithoutSponsored: Article = {
  id: 18543,
  date_gmt: '2020-11-09T09:46:40',
  title: {
    rendered: 'Sicheres Business und Inspiration für morgen: Koreas neue, einzigartige Veranstaltungsorte',
  },
  content: {
    rendered:
      '<figure id="attachment_18559" aria-describedby="caption-attachment-18559" style="width: 618px" class="wp-caption aligncenter"><img class="wp-image-18559 size-large" src="https://www.convention-net.de/wp-content/uploads/2020/11/shutterstock_335001080-w1500-h1300-1200x801.jpg" alt="" width="618" height="413" srcset="https://www.convention-net.de/wp-content/uploads/2020/11/shutterstock_335001080-w1500-h1300-1200x801.jpg 1200w, https://www.convention-net.de/wp-content/uploads/2020/11/shutterstock_335001080-w1500-h1300-600x400.jpg 600w, https://www.convention-net.de/wp-content/uploads/2020/11/shutterstock_335001080-w1500-h1300-768x513.jpg 768w, https://www.convention-net.de/wp-content/uploads/2020/11/shutterstock_335001080-w1500-h1300-110x75.jpg 110w, https://www.convention-net.de/wp-content/uploads/2020/11/shutterstock_335001080-w1500-h1300.jpg 1500w" sizes="(max-width: 618px) 100vw, 618px" /><figcaption id="caption-attachment-18559" class="wp-caption-text">Foto: Pairat Pinijkul / shutterstock.com</figcaption></figure>\n<p>Aufgrund eines exzellenten IKT-Umfelds und seiner kulturellen Attraktivität hat sich Korea als effizientes, innovatives Land für Unternehmen und Kongresse in der MICE-Branche etabliert. Hinsichtlich der Anzahl internationaler Veranstaltungen im Land belegte Korea in den letzten Jahren unter den Mitgliedern der Union der Internationalen Verbände (UIA) den ersten und zweiten Platz. Seit dem Ausbruch von Covid-19 hat die schnelle und gut organisierte Reaktion es dem Land ermöglicht, für die Sicherheit vieler Unternehmen und Konventionen zu sorgen. Mit zahlreichen Support-Programmen unterstützt Korea Verbände und Organisationen bei der Planung und Ausrichtung von Events und baut nach wie vor das Angebot an vielfältigen Eventlocations im Land aus.</p>\n',
    protected: false,
  },
  excerpt: {
    rendered:
      '<p>Aufgrund eines exzellenten IKT-Umfelds und seiner kulturellen Attraktivität hat sich Korea als effizientes, innovatives Land für Unternehmen und Kongresse in der MICE-Branche etabliert.</p>\n',
    protected: false,
  },
  featured_media: 18559,
  categories: [5, 1, 72, 170],
  acf: {
    // sponsored_by: '100HERZ',
    should_push: false,
    featured_slider: true,
    news_slider: true,
    pinned: true,
  },
  featured_image_thumb:
    'https://www.convention-net.de/wp-content/uploads/2020/11/shutterstock_335001080-w1500-h1300-300x300.jpg',
  featured_image_medium:
    'https://www.convention-net.de/wp-content/uploads/2020/11/shutterstock_335001080-w1500-h1300-600x400.jpg',
  categories_names: ['Destinationen', 'News', 'Reportagen', 'Sponsored content'],
}
