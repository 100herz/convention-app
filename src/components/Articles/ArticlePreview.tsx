import React from 'react'
import { Dimensions, Image, ImageBackground, ImageStyle, StyleSheet, View, ViewStyle } from 'react-native'
import HTML from 'react-native-render-html'
import { useNavigation } from '@react-navigation/native'
import { StackNavigationProp } from '@react-navigation/stack'
import { Ionicons } from '@expo/vector-icons'

import LogoRound from '@components/UI/LogoRound'
import Text from '@components/UI/Text'
import Touchable from '@components/UI/Touchable'
import { HomeStackParamList } from '@navigations/HomeNavigator'
import { CategoriesStackParamList } from '@navigations/CategoriesNavigator'
import { getLocaleLongDate } from '@utils/date-time'
import { Article } from '@models/article'
import { colors, defaultStyles, DefaultStyles, fonts } from '@styles/theme'

interface Props {
  article: Article
  hasImage?: boolean
  ignoreSponsored?: boolean
}

const ArticlePreview: React.FC<Props> = ({ article, hasImage = true, ignoreSponsored = false }) => {
  const navigation = useNavigation<StackNavigationProp<HomeStackParamList | CategoriesStackParamList>>()
  const isSponsored = () => !!(article.acf?.sponsored_by && article.acf?.sponsored_by.length > 0 && !ignoreSponsored)
  const isPinned = () => !!(article.acf?.pinned && !ignoreSponsored)

  const defaultArticle = (
    <Touchable onPress={() => navigation.navigate('ArticleScreen', { postId: article.id })}>
      <View style={styles.articleContainer} testID="default-article">
        {hasImage && (
          <View style={styles.imageColumn}>
            {article.featured_media ? (
              <Image style={styles.image as ImageStyle} source={{ uri: article.featured_image_thumb }} />
            ) : (
              <View style={styles.image as ViewStyle} testID="no-image">
                <LogoRound />
              </View>
            )}
          </View>
        )}
        <View style={styles.textColumn}>
          <HTML baseFontStyle={styles.title} html={article.title.rendered} />
          <View style={styles.dateContainer}>
            <Ionicons name="time" size={12} color={colors.accentColor} />
            <Text style={styles.date}>{getLocaleLongDate(new Date(article.date_gmt))}</Text>
          </View>
        </View>
      </View>
    </Touchable>
  )

  const sponsoredArticle = (
    <Touchable onPress={() => navigation.navigate('ArticleScreen', { postId: article.id })}>
      <View style={styles.sponsoredArticleContainer} testID="sponsored-article">
        <ImageBackground
          source={{ uri: article.featured_image_medium }}
          style={styles.backgroundImage}
          imageStyle={{ borderRadius: 15 }}
        >
          {isSponsored() && (
            <View style={styles.sponsoredTextContainer} testID="sponsored-text-container">
              <Text style={styles.sponsoredText}>
                Sponsored by:{' '}
                <Text style={{ ...styles.sponsoredText, fontFamily: fonts.sansBold }}>{article.acf?.sponsored_by}</Text>
              </Text>
            </View>
          )}
        </ImageBackground>
        <View style={styles.sponsoredTitleContainer}>
          <HTML baseFontStyle={styles.sponsoredTitle} html={article.title.rendered} />
        </View>
      </View>
    </Touchable>
  )

  return isSponsored() || isPinned() ? sponsoredArticle : defaultArticle
}

interface Styles extends DefaultStyles {
  articleContainer: ViewStyle
  imageColumn: ViewStyle
  image: ImageStyle | ViewStyle
  textColumn: ViewStyle
}

const styles = StyleSheet.create<Styles>({
  ...defaultStyles,
  backgroundImage: {
    ...defaultStyles.backgroundImage,
    height: Dimensions.get('screen').width - 100,
  },
  articleContainer: {
    flexDirection: 'row',
    padding: 15,
  },
  imageColumn: {
    width: 100,
    height: 100,
    justifyContent: 'center',
    borderRadius: 15,
  },
  image: {
    flex: 1,
    borderRadius: 15,
  },
  textColumn: {
    width: Dimensions.get('screen').width - 130,
    paddingHorizontal: 15,
    alignItems: 'flex-start',
  },
  title: {
    fontFamily: fonts.sans,
    fontSize: 13,
    lineHeight: 13 * 1.25,
    color: colors.primaryColor,
  },
})

export default ArticlePreview
