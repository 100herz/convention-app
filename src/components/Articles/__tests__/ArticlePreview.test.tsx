import React from 'react'
import { act, fireEvent, render } from '@testing-library/react-native'

import ArticlePreview from '../ArticlePreview'
import { article, articleWithoutFeatured, articleWithoutSponsored } from '@__mocks__/article'

export const mockedNavigate = jest.fn()
jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native')
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockedNavigate,
    }),
    useNavigationParam: jest.fn(jest.requireActual('@react-navigation/native').useNavigationParam),
  }
})

describe('<ArticleList />', () => {
  it('renders correctly', () => {
    const { toJSON } = render(<ArticlePreview article={article} />)
    expect(toJSON()).toMatchSnapshot()
  })

  it('has no image and renders a View with three texts instead', () => {
    const { getByTestId } = render(<ArticlePreview article={articleWithoutFeatured} />)
    expect(getByTestId('no-image').children).toHaveLength(3)
  })

  it('should call the navigate function once', () => {
    const { getByTestId } = render(<ArticlePreview article={articleWithoutSponsored} />)

    act(() => {
      fireEvent.press(getByTestId('button'))
    })

    expect(mockedNavigate).toHaveBeenCalledTimes(1)
  })
})
