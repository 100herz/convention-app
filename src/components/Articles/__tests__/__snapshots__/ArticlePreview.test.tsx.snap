// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArticleList /> renders correctly 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <View
    style={
      Object {
        "borderRadius": 10,
        "height": 740,
        "padding": 10,
      }
    }
  >
    <View
      accessibilityIgnoresInvertColors={true}
      style={
        Object {
          "flex": 1,
          "justifyContent": "space-between",
          "resizeMode": "cover",
        }
      }
    >
      <Image
        source={
          Object {
            "uri": "https://www.convention-net.de/wp-content/uploads/2020/11/iStock-1250329916-Personen_mit_Masken-w1500-h1300-300x300.jpg",
          }
        }
        style={
          Array [
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            },
            Object {
              "height": undefined,
              "width": undefined,
            },
            Object {
              "borderRadius": 15,
            },
          ]
        }
      />
      <View
        style={
          Object {
            "alignItems": "flex-end",
            "backgroundColor": "rgba(223, 227, 232, 0.7)",
            "borderTopLeftRadius": 15,
            "borderTopRightRadius": 15,
            "paddingHorizontal": 30,
            "paddingVertical": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#333333",
              "fontFamily": "sans",
              "fontSize": 14,
            }
          }
        >
          Sponsored by 
          <Text
            style={
              Object {
                "color": "#333333",
                "fontFamily": "sans-bold",
              }
            }
          >
            XYZ
          </Text>
        </Text>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "rgba(223, 227, 232, 0.7)",
            "borderBottomLeftRadius": 15,
            "borderBottomRightRadius": 15,
            "justifyContent": "flex-end",
            "paddingHorizontal": 30,
            "paddingVertical": 10,
          }
        }
      >
        <View
          style={Object {}}
        >
          <Text
            allowFontScaling={true}
            selectable={false}
            style={
              Array [
                Object {},
                Object {},
              ]
            }
          >
            <Text
              allowFontScaling={true}
              style={
                Object {
                  "fontFamily": "sans",
                  "fontSize": 16,
                }
              }
            >
              „New Normal“ bei Leonardo Hotels – mit neuen MICE-Formaten und Konzepten
            </Text>
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
